clang/lib/Sema/SemaStmt.cpp

```diff
@@ -4418,3 +4418,57 @@ StmtResult Sema::ActOnCapturedRegionEnd(Stmt *S) {
 
   return Res;
 }
+
+WhereClause *Sema::ActOnWhereClause(SourceLocation Loc) {
+  return new (Context) WhereClause(Loc);
+}
+
+BoundsFact *Sema::ActOnBoundsFact(IdentifierInfo *Id, Expr *E,
+                                  Scope *CurScope, SourceLocation Loc) {
+  BoundsExpr *Bounds = dyn_cast<BoundsExpr>(E);
+  if (!Bounds)
+    return nullptr;
+
+  LookupResult Lookup(*this, Id, Loc, Sema::LookupOrdinaryName);
+  LookupParsedName(Lookup, CurScope, nullptr, true);
+  if (Lookup.empty()) {
+    Diag(Loc, diag::err_undeclared_var_use) << Id->getName();
+    return nullptr;
+  }
+
+  VarDecl *VD = Lookup.getAsSingle<VarDecl>();
+  if (!VD)
+    return nullptr;
+
+  return new (Context) BoundsFact(VD, Bounds, Loc);
+}
+
+EqualityOpFact *Sema::ActOnEqualityOpFact(Expr *E, SourceLocation Loc) {
+  // We define an equality-op fact in terms of equality-expressions as defined
+  // in section 6.5.9 of the C11 spec. Equality-op facts have an added
+  // constraint that the equality-expressions should be non-modifying
+  // expressions
+
+  // equality-expression:
+  //   relational-expression
+  //   equality-expression == equality-expression
+  //   equality-expression != equality-expression
+
+  Lexicographic Lex(Context, nullptr);
+  Expr *TmpE = Lex.IgnoreValuePreservingOperations(Context, E);
+
+  // TODO: Handle equality-op facts joined by logical operators, like
+  // _Where equality-op-fact && equality-op-fact || equality-op-fact.
+
+  auto *BO = dyn_cast<BinaryOperator>(TmpE);
+  // Note: isComparisonOp is a function which checks for equality and
+  // relational operators.
+  if (!BO || !BO->isComparisonOp())
+    return nullptr;
+
+  if (!CheckIsNonModifying(BO->getLHS()) ||
+      !CheckIsNonModifying(BO->getRHS()))
+    return nullptr;
```

Comment:
Does CheckIsNonModifying emit an error message?

Reply:
AFAIK and I checked, `CheckIsNonModifying` does not issue an error message.